### ISSUE TITLE
Bug: `promise-function-async` is reporting on `() => A | Promise<B>`

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#10687]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.ts&code=FAMwrgdgxgLglgewgAgEYAoCUyDexnIBOApjGISgLICGMAFgHSHUQAmCAtlsgPzIDkSYv2QAuZAAVCnOAGdiTYrIQAbAG7F0ARkwBuYAF8gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Y6AewFtLlFaAM1hMy%2BSpya0AhskKj0URNC7RI4MAF8QmoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
function b() {
  return Math.random() ? 'one' : Promise.resolve(1);
}
```

### ESLint Config

```javascript
_playground default_
```

### tsconfig

```jsonc
_playground default_
```

### Expected Result

No reports

### Actual Result

"Functions that return promises must be async."

Attempts to fix to `async function`, changing the return type from `A | Promise<B>` to `Promise<A | B>`

### Additional Info

While this behavior is [described in the docs/demo](https://typescript-eslint.io/rules/promise-function-async/), I believe it's wrong because the function _does not_ "return a promise," it only "MAY return a promise." 

**The change requested is unsafe** as it changes the signature/type of the function entirely.

Note that whether `() => A | Promise<B>` is a good type is not relevant to this issue. The type exists in the wild ([web extensions](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/runtime/onMessage), [screenshot](https://github.com/user-attachments/assets/91c93fc8-a5cf-43e8-85b1-fedcea0b4870)) and it can be prohibited by a separate rule ([suggested here](https://github.com/typescript-eslint/typescript-eslint/issues/9296))

**Comment by bradzacher** [comment]
As you mentioned - the rule docs explicitly state this and they also include a workaround if it is your intention to return a non-promise and a promise:

> If it is intentional, make the return type explicitly to allow the rule to pass.


---

> The change requested is unsafe as it changes the signature/type of the function entirely. 

The change requested is a suggestion fixer which must be manually applied by a user. Suggestion fixers (unlike autofixers) are allowed to be unsafe or incomplete as they must be manually actioned.

**Comment by fregante** [comment]
> If it is intentional, make the return type explicitly to allow the rule to pass.

But this is not what the error says: "Functions that return promises must be async."

The error message, as is, is incomplete/incorrect because the function does not return only a promise.

The reason why I opened this issue is specifically because I found an `eslint-disable promise-function-async` explanation that missed this subtlety. And honestly, even if I understand the reason for this exclusion (i.e. DX), I think it shouldn't exist.

Changes suggested, one or more:

1. improve error message **_for this case_**, e.g. "Functions that return promises must be async or have an explicit return type"
2. remove this exception, report regardless of return type presence (because I don’t think inference should validated or invalidate this rule)
2. extract `A | Promise<B>` reporting to its own rule, e.g. https://github.com/typescript-eslint/typescript-eslint/issues/9296
3. stop reporting on functions that return non-promises (because, once again, this function does not return _only_ a promise, so it should not be reported)

**Comment by bradzacher** [comment]
> remove this exception, report regardless of return type presence
> because I don’t think inference should validate or invalidate this rule

In the vast, vast, vast majority of cases the developer intent is to just return a promise and not to have a hybrid return type - and if you asyncify the function then you get that exact behaviour - so the rule reporting is correct and guides the user towards better code.

But there are cases where you needs to intentionally NOT asyncify the function because you NEED a hybrid output. Instead of just relying on a disable comment here we chose to take the return type as explicit developer intent that a hybrid return type was desired.

So put another way:
1. no explicit return type + hybrid return type = highly likely wanted just promise
2. explicit return type + hybrid return type = definitely wanted a hybrid return type
3. promise return type (with or without an explicit return type) = definitely wanted just promise

So the rule reports in (1) and (3) and not in (2).

**Comment by fregante** [comment]
> * no explicit return type + hybrid return type = highly likely wanted just promise

Right, but that's what I meant in my last point, the rule is broad and it enforces something that is not in the explicit intent of the rule.

Even if you don't agree to change that, suggestions 1 and 3 still apply (error message + additional fixer)

> Instead of just relying on a disable comment 

Yeah, that's why it would be great to update the error message to suggest this. Most developers barely read the error messages, let alone googling the rule name to read its docs.

**Comment by bradzacher** [comment]
> stop reporting on functions that return non-promises
> because, once again, this function does not return only a promise, so it should not be reported 

Again I want to stress that in the vast, vast, vast majority of cases a user **_DOES NOT WANT A HYBRID RETURN TYPE_** and instead the hybrid return type is an inference error caused by a function that returns a plain and a promise value.

Eg cases like this:

```ts
function doSomethingAsync(): Promise<number>;

function doSomething(condition: boolean) {
  if (condition) {
    return 1;
  }
  return doSomethingAsync();
}
```

Generally the intent here is "this function returns `Promise<number>`" because you generally can't pass `number | Promise<number>` anywhere and you can't call promise methods on it. And thus generally speaking users want the Promise return type.

**Comment by bradzacher** [comment]
My 2c is that
- this is and "always"[1] has been the intended behaviour for many, many years
- most users expect this case to be reported because most usecases don't want a hybrid return type.

So removing the behaviour goes against that established and expected practice.

I'd be okay with us updating the error message if the function has a hybrid return type to mention that an explicit annotation will silence the rule.

----


[1] iirc this case was added after the rule was first created but regardless it's near enough to "always".
